pkg/controller/elasticsearch/driver/nodes.go

```diff
@@ -33,23 +32,32 @@ func (d *defaultDriver) reconcileNodeSpecs(
 		return results.WithError(err)
 	}
 
-	if !d.Expectations.GenerationExpected(actualStatefulSets.ObjectMetas()...) {
-		// Our cache of StatefulSets is out of date compared to previous reconciliation operations.
-		// Continuing with the reconciliation at this point may lead to:
-		// - errors on rejected sset updates (conflict since cached resource out of date): that's ok
-		// - calling ES orchestration settings (zen1/zen2/allocation excludes) with wrong assumptions: that's not ok
-		// Hence we choose to abort the reconciliation early: will run again later with an updated cache.
-		log.V(1).Info("StatefulSet cache out-of-date, re-queueing", "namespace", d.ES.Namespace, "es_name", d.ES.Name)
+	// check if actual StatefulSets and corresponding pods match our expectations before applying any change
+	ok, err := d.expectationsMet(actualStatefulSets)
+	if err != nil {
+		return results.WithError(err)
+	}
+	if !ok {
 		return results.WithResult(defaultRequeue)
 	}
 
-	expectedResources, err := expectedResources(d.Client, d.ES, observedState, keystoreResources)
+	expectedResources, err := nodespec.BuildExpectedResources(d.ES, keystoreResources)
 	if err != nil {
 		return results.WithError(err)
 	}
```

Comment:
So after this state we may have updated some ssets, so we'll end up with sset update conflicts in next steps (downscale, rolling upgrade). Which is fine, but I'm wondering whether we should maybe either:
* requeue here if the downscale step did any change
* re-check expectations here to catch any change

Reply:
We may also recheck expectations before `downscale` and `handleRollingUpgrades` It would make it explicit and maybe less error prone when this code will be updated.